### ISSUE TITLE
New devices on host are not exposed to privileged container

### ISSUE THREAD

**Issue by grepory** [docker/docker#16160]
I have a very simple container, basically gliderlabs/alpine:3.2 with python running on a machine in AWS.

Inside that container, I have a script that attaches an EBS volume to the instance and then mounts it. To run the container I am using:

docker run --privileged -v /mnt:/mnt

I then run the script. It attaches the EBS volume to /dev/xvdc, but when it goes to mount the volume, /dev/xvdc isn't there.

```
subprocess.CalledProcessError: Command '['mount', '/dev/xvdc', '/mnt/nsqd-1']' returned non-zero exit status 255
bash-4.3# mount /dev/xvdc /mnt/nsqd-1
mount: mounting /dev/xvdc on /mnt/nsqd-1 failed: No such file or directory
bash-4.3# ls /mnt
nsqd-1
```

Outside the container, the device exists:

```
core@ip-172-31-8-49 ~ $ ls /dev/xvdc
/dev/xvdc
```

If a new device is exposed in /dev, I would expect to see it exposed to a privileged container also.

**Comment by phemmer** [comment]
For the docker maintainers: is there any reason not to use `devtmpfs` (instead of the current vanilla `tmpfs`) for `/dev` inside a privileged container? That would solve this issue.

@grepory as a workaround, you could `-v /dev/disk:/dev/disk`. This wouldn't give you `sdX`/`xvdX` labels, but you would get things like `by-uuid` and `by-id` instead. I would argue that these are much better to use anyway, as `sdX`/`xvdX` is unreliable since it depends on the order disks were added/detected.

**Comment by coolljt0725** [comment]
Maybe `mount -t devtmpfs xx /dev` in the container is a way, in this way
you can find the new device in host. don't know if this is a proper way :-)

**Comment by phemmer** [comment]
> Maybe `mount -t devtmpfs xx /dev` in the container is a way

Unfortunately not. There are mounts within `/dev` (`/dev/shm`,`/dev/pts`, etc), and doing so would make them inaccessible.

**Comment by grepory** [comment]
So what I'm currently doing is this, and it feels hella jank to me, but hey Get Things Done, right?

```python
def ensure_filesystem(device, svcname):
  if not os.path.exists(device):
    devname = device.split('/')[-1]
    major, minor = os.readlink('/sys/block/%s/bdi' % (devname),).split('/')[-1].split(':')
    subprocess.check_call(['mknod', device, 'b', major, minor])

  blkid = subprocess.check_output(['blkid', device])
  if not re.search('.*TYPE="ext4".*', blkid):
    subprocess.check_call(['mkfs.ext4', '-L %s' % (svcname,), device])

def get_next_free_block_device():
  # Skip a, b because reserved for instance and ephemeral storage.
  alldisks = sets.Set(["/dev/xvd%s" % (x,) for x in string.ascii_lowercase[2:15]])
  mounts = subprocess.check_output(['cat', '/proc/mounts']).split('\n')
  mounted = sets.Set()

  regex = re.compile('^(/dev/xvd.).*')
  for line in mounts:
    m = re.match(regex, line)
    if m:
      mounted |= sets.Set([m.group(1)])
  print "Mounted disks: %s" % (' '.join(mounted),)
  return (alldisks - mounted).pop()

def attach_volumes(svcname):
  metadata = get_instance_metadata(timeout=2)
  instance_id = metadata['instance-id']
  for v in get_service_volumes(svcname):
    dev = get_next_free_block_device()
    print "Mounting to %s" % (dev,)

    resp = v.attach_to_instance(
      InstanceId=instance_id,
      Device=dev,
    )

    mount_path = "/mnt/%s" % (svcname,)
    mkdir_p(mount_path)
    ensure_filesystem(dev, svcname)
  subprocess.check_call(['udevadm', 'trigger'])
  subprocess.check_call(['udevadm', 'settle'])
  subprocess.check_call(['partprobe'])
```

Fortunately I have good access to the sysfs, which lets me find the block device identifier so that I can then run `mknod`.

So I just create the new device in /dev if it doesn't exist, and then add a label when I `mkfs.ext4` on it (if there is no existing filesystem).

THEN, I can mount it from outside of the container because the label is something programmatically determined by the service name. (This is all so that I can use EBS volumes and have them follow individual units around in my CoreOS Fleet).

The udevadm triggers and partprobe work only because this is a privileged container, it seems. It just feels jank, but it works.

**Comment by grepory** [comment]
Oh that's another thing. I have to mount it from outside the container because can't mount. That feels weird, and I don't understand why that doesn't work--but I'm okay with it. If someone can explain why that doesn't work because they just happen to know--that would be awesome. <3

**Comment by phemmer** [comment]
@grepory see #10088 #14630 #14097 #15648

**Comment by grepory** [comment]
@phemmer Thanks so much -- you are freaking amazing.

**Comment by LK4D4** [comment]
This is still a real issue, not sure if we can solve it or let advanced users solve it themselves.

**Comment by Silex** [comment]
What is the current way of correcting this issue? I run a container where USB keys gets plugged in the host and it only "works" when the USB slots are filled prior to running the container (`/dev/sdb`, `/dev/sdc` already exist).

The `docker run` documentation is unclear about how `:shared` or `:slave` solve that issue.

**Comment by fabiand** [comment]
I'm solving this by bind-mounting /dev into the container plus some additional hacks,

I'm not aware of the implementation, but do me `/dev/` inside the container looks like a manually assembled `/dev/, and not inherited from the hosts `/dev/` path, if that is the case, then `:shared` and `:slave` will not work

**Comment by Silex** [comment]
@fabiand: yes, I was about to comment that `-v /dev:/dev --privileged` seems to work.

Thanks!

**Comment by fabiand** [comment]
@Silex well - I've seen a few issues around terminals when using plain `-v /dev:/dev`, this required me to keep the _old_ `/dev/pts` and `/dev/ptmx`. Besides that I noticed that `/dev/mqueue` and `/dev/shm` seem to be special (and also kept the old versions).

**Comment by Silex** [comment]
@fabiand: my guess is that there is a cleaner way that involves udev, mknod & friends, but I don't have time to investiguate atm.

**Comment by fabiand** [comment]
Not sure if using udev inside the container is a good idea.
After all, we already have a udev on the host side. You might run into trouble if you i.e. got different udev rules inside the container than on the host. This could lead to different `/dev` trees.

To me it looks liek we should just have one `/dev` per host, just a nice way to present this inside a container is missing.

Just my 2ct.

**Comment by ejstrates** [comment]
Are the above workarounds only with respect to running Docker on a Linux host?

Using Docker for Mac, I don't seem to be able to mount devices using -v, --privileged, or --device.

I also tried restarting Docker for Mac after the device had been attached, to no avail.

**Comment by ejstrates** [comment]
wrt Docker for Mac; this is an open issue: docker/for-mac#900

**Comment by princerachit** [comment]
2018 and still facing this issue. Any plans to work on this?

**Comment by Silex** [comment]
Somewhat duplicate of https://github.com/moby/moby/issues/27886

**Comment by bclouser** [comment]
I reverberate @princerachit's sentiment. Also knowing of an official workaround, or best practices would be nice.

**Comment by akhilerm** [comment]
Any official workaround for this issue yet??

**Comment by DuBistKomisch** [comment]
`/dev:/dev` as a volume worked well enough for me
